### ISSUE TITLE
Cannot find info on hosting terminal in another app.

### ISSUE THREAD

**Issue by sharpninja** [microsoft/terminal#4572]
We need documentation on hosting the console.  Project templates would be very helpful.

**Comment by DHowett-MSFT** [comment]
This is very well-documented [on our docs.microsoft.com section](https://docs.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session)

**Comment by DHowett-MSFT** [comment]
There's also samples [in this repository](https://github.com/microsoft/terminal/tree/master/samples/ConPTY).

**Comment by sharpninja** [comment]
There are no samples or document for hosting in a UWP process, which has been promised for a long time.  Please reopen.

**Comment by DHowett-MSFT** [comment]
We haven’t promised documentation on that, because hosting an unconstrained application from a UWP context is currently impossible. This is why we didn’t build the terminal as a UWP. That’s just not something the platform supports.

**Comment by sharpninja** [comment]
This is literally line 4, as a Priority 0 item:

> Terminal's core will be hostable as a UWP (and perhaps WPF) Control so that apps can host/embed a high quality Terminal. This will satisfy a long-standing ask from many customers and partners for a hostable/embeddable Terminal Control.

[source](https://github.com/microsoft/terminal/blob/master/doc/terminal-v1-roadmap.md)
________________________________

**Comment by zadjii-msft** [comment]
Okay you're not wrong about that. That doc is maybe a little out of date, but I guess we did mention that publicly.


The way the terminal's architecture is designed, it's certainly _possible_ to re-use the terminal control in other UWP projects. However, I think we've let the priority of that particular feature slip a bit as we approach 1.0. AS it currently stands, I don't think that we'll have the cycles to be able to _officially_ support  a nuget package or something for the terminal control in the 1.0 timeframe. I think it was more important for us to _prepare_ for allowing arbitrary apps to host the control, and that we've most certainly done.

Thant being said, it'd definitely be possible to _unofficially_ use the `TerminalControl.dll` and `TerminalControl.winmd` that's built from this project in other projects. I'm not saying you _should_ do this. I'm saying you _could_. 

@DHowett-MSFT is alluding to the fact that launching other processes from a UWP context is fraught with peril. Those perils is why we've chosen the Win32+Xaml Hosting model for the Terminal. That being said, you could totally add a terminal to a pure UWP application. Just be aware that it's a _here be dragons_ path.

We probably should add documentation on this, but I'm going to toss it on the backlog, since I don't think we'll want to author documentation at this time while things are still changing pretty rapidly.

**Comment by DHowett-MSFT** [comment]
Ah, sorry about that!

**Comment by sharpninja** [comment]
# Totally cool!  

I am very happy to _guinea pig this_ as I'm working on a project that literally is providing a programmer's "toolbox" that having an embedded terminal is a foundational requirement, and I started this project after months ago reading that quote and anxiously awaiting the release of the hostable console.  I understand not wanting to post any formal docs yet, but if you could at least give me a rough set of steps to get a terminal hosted in a C# UWP app then I'd be greatly appreciative.

**Comment by DHowett-MSFT** [comment]
@sharpninja If you're willing to take instructions with _no warranty whatsoever_ and do something that might simply break for no reason because we're not ready to ship it yet, read on...

<details>
<summary>And on and on</summary>
If you _build Terminal_, you'll get a bunch of interesting things in `./x64/Debug`:

* `TerminalControl.dll` and `TerminalControl.winmd`: These implement the terminal control; the winmd is how you'll import references to another project. This is _easier_ with C#, actually, but I don't know how it's going to feel about activating classes from these DLLs.
* `resources.pri`: Localizable resources from the entire terminal app and all of microsoft.ui.xaml (sorry: we can't split it finer than this right now; you might be able to grab a `pri` from x64/Debug/TerminalControl?)
* `TerminalSettings.dll` and `TerminalSettings.winmd`: A single-class library that lets you construct an object that can be used to pass settings to a terminal.
* `TerminalConnection.dll` and `TerminalConnection.winmd`: Contains the ITerminalConnection interface you'll need to conform to _and_ a couple other classes. If you don't need the conpty, azure or telnet connections you should be able to get by with only the winmd? If you copy the dll, make sure to copy `cpprest` and `telnetpp` DLLs as well.

Know that you'll be going significantly off-road here.

Anyway, once you grab those things and integrate them into your own package, you should be able to host a `Microsoft.Terminal.TerminalControl.TermControl` (yes, that is a heck of a name) in your Xaml tree. It can't be constructed without an `IControlSettings` (TerminalSettings winmd) and an `ITerminalConnection` (TerminalConnection winmd).

There might be one or two .xbf files you need as well; if you see it failing to load SearchBoxControl, you'll want to make sure you copy that over too.

It's really not in a good place right now!
</details>

**Comment by sharpninja** [comment]
Thank you so much!  I know what I’m doing this weekend.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by DHowett-MSFT** [comment]
If you want to go significantly the _other_ way on this, the WpfTerminalControl project will produce a nuget package with far fewer dependencies that'll work with _WPF_ Xaml (yeah!) down to Windows 7ish. I know that's really outside the scope of UWP :wink:

**Comment by RWeigelt** [comment]
> If you want to go significantly the _other_ way on this, the WpfTerminalControl project will produce a nuget package with far fewer dependencies that'll work with _WPF_ Xaml [...]
Where can I find more information about WpfTerminalControl and how to use it in my WPF application?

**Comment by DHowett** [comment]
Sorry -- it's undocumented for a reason. It's very much an act of trailblazing to get it going.

**Comment by akasarto** [comment]
Hey! If someone is still interested in this, I've been playing around with the available [WpfTerminalControl](/microsoft/terminal/tree/main/src/cascadia/WpfTerminalControl) and the simplest way I found to make this work is as follows:

- Make sure the WPF app has the compiled `PublicTerminalCore.dll` available for _DLLImport/PInvoke_. I was able to achieve this by including the _microsoft/terminal_ repository as a submodule of mine and creating a script called `build-deps.cmd` to compile only the necessary bits: `/modules/ms-terminal/src/cascadia/PublicTerminalCore`.
- Add a reference to the submodule `/src/cascadia/WpfTerminalControl` project to your solution.
- Implement the `ITerminalConnection` to handle user input.
- On your WPF app, add a reference to the `WpfTerminalControl` project above and instantiate the user control `TerminalControl.xaml`, assigning your custom `IterminalConnection` implementation to it.

**Note that for the compilation script above to work, the `Prerequisites` from the [microsoft/terminal](/microsoft/terminal) must be met.

The implementation referred above can be found under the `net5-refactor` branch, [here](https://github.com/akasarto/d-term/tree/net5-refactor)

For my use case, the result looks like this:

![dTerm](/akasarto/d-term/blob/net5-refactor/media/dTerm.png?raw=true "dTerm")

**Comment by Crypties** [comment]
> It's really not in a good place right now!

About 3 years later. Is it in a good place now?

**Comment by zadjii-msft** [comment]
Not really. This unfortunately hasn't really bubbled up to the top of our list of priorities yet. 

* The WPF control is _technically_ hostable in another app, but not in a productized sense. See #15053. You could surely follow that path, with the caveat that _the entire interface is subject to change_. 
* Speaking of productization, we're mostly tracking that in #6999. 
* We really want to refine the whole HwndTerminal / TerminalCore / ControlCore/Interactivity story. There's a lot of duplicated work there, and we could surely just replace one with the other.
